query/src/storages/fuse/operations/delete_from.rs

```diff
@@ -0,0 +1,155 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+use std::sync::Arc;
+
+use common_datablocks::DataBlock;
+use common_datavalues::DataSchemaRef;
+use common_datavalues::DataSchemaRefExt;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_planners::DeletePlan;
+use common_planners::Expression;
+
+use crate::pipelines::transforms::ExpressionExecutor;
+use crate::sessions::QueryContext;
+use crate::storages::fuse::io::BlockReader;
+use crate::storages::fuse::io::MetaReaders;
+use crate::storages::fuse::meta::BlockMeta;
+use crate::storages::fuse::meta::SegmentInfo;
+use crate::storages::fuse::meta::TableSnapshot;
+use crate::storages::fuse::pruning::BlockPruner;
+use crate::storages::fuse::pruning::Pred;
+use crate::storages::fuse::FuseTable;
+use crate::storages::index::BlockStatistics;
+use crate::storages::index::RangeFilter;
+
+impl FuseTable {
+    pub async fn do_delete_from(&self, ctx: Arc<QueryContext>, plan: DeletePlan) -> Result<()> {
+        if let Some(expr) = plan.selection {
+            let mut segments_kept: Vec<Arc<SegmentInfo>> = vec![];
+            let mut segments_new: Vec<SegmentInfo> = vec![];
+
+            if let Some(snapshot) = self.read_table_snapshot(ctx.as_ref()).await? {
+                let segment_locs = snapshot.segments.clone();
+
+                if segment_locs.is_empty() {
+                    return Ok(());
+                } else {
+                    let schema = self.table_info.schema();
+                    let block_pred: Pred = {
+                        let verifiable_expression = RangeFilter::try_create(&expr, schema)?;
+                        Box::new(move |v: &BlockStatistics| verifiable_expression.eval(v))
+                    };
+
+                    for loc in segment_locs {
+                        let reader = MetaReaders::segment_info_reader(&ctx);
+                        let segment_info = reader.read(loc).await?;
+                        let blocks =
+                            BlockPruner::filter_segment(segment_info.as_ref(), &block_pred)?;
+                        if blocks.is_empty() {
+                            // the segment and the all the blocks should be kept definitely
+                            segments_kept.push(segment_info);
+                        } else {
+                            // Maybe there are rows should be deleted (false-positive of pruner)
+                            // use `delete_rows` to do the real data filtering.
+                            let new_segment = self.delete_rows(ctx.clone(), blocks, &expr).await?;
+                            segments_new.push(new_segment);
+                        }
+                    }
+
+                    let new_snapshot = self.merge_segments(&segments_kept, &segments_new);
+                    self.commit_deletion(new_snapshot).await?
```

Comment:
So `deletion` will block `inserts` & `updates`?

Reply:
`inserts` does not block other mutations. In this pr, a rather simple rule will be applied: when there are write-write conflicts,  first-committer-win.

Reply:
So continuous small inserts will cause the heavy deletion retry ...

Reply:
> So continuous small inserts will cause the heavy deletion retry ...

er... too awkward... to catch this message months later. 

- yes, OCC implies the possibility of retrying
- deletion might be heavy, but the retrying of deletion might not be